### PR TITLE
New style GEDelectron and GEDphoton regressions to EGM reco sequence in 90X

### PR DESCRIPTION
**Pull request by rafaellopesdesa**
PR for new style of EGM energy correction and per-particle resolution regression. It is similar to PR #16968 which was intended for standalone application by users (and that is still under review). This one modifies the EGM reco sequence, so changes are expected in the validation plots, specially in plots related to electron and photon energy. This PR depends on the PR #17048, which I believe has already been merged, with the new name convention for the regressions in the GlobalTag. I've tested locally in details with workflow 250200.0 and the results were good.

Att: @fcouderc

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 1]
```diff
@@ -82,26 +85,31 @@ class EGExtraInfoModifierFromDB : public ModifyObjectValueBase {
...
   std::vector<const GBRForestD*> e_forestH_mean_;
-  std::vector<const GBRForestD*> e_forestH_sigma_; 
-  const GBRForest* ep_forestH_weight_;
+  std::vector<const GBRForestD*> e_forestH_sigma_;
+
+  const CaloTopology *topology_;
+
+  bool useLocalFile_;
+  std::string addressLocalFile_;  
+  TFile* pointerLocalFile_;
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: additional open files during execution should be avoided as much as possible.
Open the file, read what's needed from it into memory and then close the file.
Do this if memory profile allows.

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 2]
```diff
@@ -111,22 +119,20 @@ DEFINE_EDM_PLUGIN(ModifyObjectValueFactory,
...
-  vtxTag_ = conf.getParameter<edm::InputTag>("vertexCollection");
-  
-  if (autoDetectBunchSpacing_) {
-    bunchspacingTag_ = conf.getParameter<edm::InputTag>("bunchSpacingTag");
-  } else {
-    bunchspacing_ = conf.getParameter<int>("manualBunchSpacing");
+  useLocalFile_     = conf.getParameter<bool>("useLocalFile");
+  if (useLocalFile_) {
+    addressLocalFile_ = conf.getParameter<std::string>("addressLocalFile");
+    pointerLocalFile_ = TFile::Open(addressLocalFile_.c_str());
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: use FileInPath

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 3]
```diff
@@ -364,118 +399,126 @@ namespace {
...
   if( missing_clusters ) return ; // do not apply corrections in case of missing info (slimmed MiniAOD electrons)
+
+  const bool iseb = ele.isEB();  
+  edm::Handle<EcalRecHitCollection > ecalRecHits;
+  if (iseb) ecalRecHits = ecalRecHitsEB_ ;
+  else      ecalRecHits = ecalRecHitsEE_ ;
+  if ( !ecalRecHits.isValid() ) return;
   
-  std::array<float, 33> eval;  
+  Int_t N_SATURATEDXTALS  = 0;
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: please use base C++ types.
Also, all caps for a variable is a bad style, it's usually used for compiler macros and some constants.

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 4]
```diff
@@ -364,118 +399,126 @@ namespace {
...
-  eval[14] = ess.sigmaIphiIphi;
-  eval[15] = std::max(0,numberOfClusters-1);
-  
+  const double raw_es_energy = the_sc->preshowerEnergy();
+  const auto& full5x5_ess = ele.full5x5_showerShape();
+
+  eval[0]  = raw_energy;
+  eval[1]  = the_sc->etaWidth();
+  eval[2]  = the_sc->phiWidth(); 
+  eval[3]  = full5x5_ess.e5x5/raw_energy;
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: check for full5x5_ess.e5x5 to be non-zero and set all values with potential divisions by zero to something sensible below. (maybe define ``` e5x5Inverse = full5x5_ess.e5x5 != 0 ? 1.f/full5x5_ess.e5x5 : 0 ``` )

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 5]
```diff
@@ -484,189 +527,236 @@ void EGExtraInfoModifierFromDB::modifyObject(reco::GsfElectron& ele) const {
...
+  auto el_track = ele.gsfTrack();
+
+  const float trkMomentum = el_track->pMode();
+  const float trkEta      = el_track->etaMode();
+  const float trkPhi      = el_track->phiMode();
+  
+  float ptMode       = el_track->ptMode();
+  float ptModeErrror = el_track->ptModeError();
+  float etaModeError = el_track->etaModeError();
+  float pModeError   = sqrt(ptModeErrror*ptModeErrror*cosh(trkEta)*cosh(trkEta) + ptMode*ptMode*sinh(trkEta)*sinh(trkEta)*etaModeError*etaModeError);
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: Can you use qoverpModeError instead of this rather unnatural value?

File: RecoEgamma/EgammaTools/python/regressionWeights_cfi.py [hunk 6]
```diff
@@ -1,29 +1,82 @@
-from CondCore.DBCommon.CondDBSetup_cfi import *
+import FWCore.ParameterSet.Config as cms
+
+def regressionWeights(process):
```
Comment (slava77) [inline RecoEgamma/EgammaTools/python/regressionWeights_cfi.py]: why is this file needed if you already have the GT updated in the release?

**Comment by rafaellopesdesa** [comment]
@slava77 I will apply the corrections in the code review. I have some comments about the other questions here:

   * I tried to run the workflows that failed locally. They failed because the regressions cannot be found in the mcRun1 GlobalTags (I ran adding the regressions by hand). I have contacted the AlCa conveners to check why the mcRun2 queues were updated but not the mcRun1 ones.
   * Yes, we could have used a more natural definition of the uncertainty in the track momentum. In particular error on qOverp is an excellent suggestion. However, the training has already be done with this (rather convoluted, I agree) definition. We will keep your suggestion in mind for the next round.
   * regressionWeights_cfi.py is not necessary. I just updated it for "historical reasons". It was there before and I updated it pointing to the new names (which supposedly are in the GT). If preferred, I can remove this configuration from RecoEgamma.

Should I go updating the code with your review or should I wait the confirmation from AlCa that the mcRun1 queues have been committed as well before doing that?

**Comment by slava77** [comment]
On 12/28/16 12:43 PM, Rafael Lopes de Sa wrote:
> @slava77 <https://github.com/slava77> I will apply the corrections in
> the code review. I have some comments about the other questions here:
> 
>   * I tried to run the workflows that failed locally. They failed
>     because the regressions cannot be found in the mcRun1 GlobalTags (I
>     ran adding the regressions by hand). I have contacted the AlCa
>     conveners to check why the mcRun2 queues were updated but not the
>     mcRun1 ones.

OK.
Assuming the payloads are available, it should be straightforward to get
the updates in.

>   * Yes, we could have used a more natural definition of the uncertainty
>     in the track momentum. In particular error on qOverp is an excellent
>     suggestion. However, the training has already be done with this
>     (rather convoluted, I agree) definition. We will keep your
>     suggestion in mind for the next round.

Can you check the math with a few printouts for what goes into the
training inputs
using qOverP and this pError.
It could be that the final value that's computed will be the same, to
within numerical precision.
The problem is that you are making a numerical roundtrip from qOverP to
pt and eta and then back to p, which is slow and can be a source of bad
values for no good reason.


>   * regressionWeights_cfi.py is not necessary. I just updated it for
>     "historical reasons". It was there before and I updated it pointing
>     to the new names (which supposedly are in the GT). If preferred, I
>     can remove this configuration from RecoEgamma.

Please remove it.

> 
> Should I go updating the code with your review or should I wait the
> confirmation from AlCa that the mcRun1 queues have been committed as
> well before doing that?

Both will be necessary to proceed with this PR.
It's up to you to wait or update the code now.


Is it correct to assume that you will not need to support the old
training with the old variables in 90X?

Do we now have a copy of the new training processing code here and in
EgammaAnalysis?
If so, the copy in the analysis better be removed to reduce possible
confusion when someone just keeps using 80X recipes on 91X.

**Comment by rafaellopesdesa** [comment]
Hi @slava77,

I've updated the codes with your recommendations. Some further comments below:

   * Yes, using the qOverp error returned the same numerical value (to numerical precision). I changed as you recommended.
   * You are correct that the old regression style will *not* be supported in 90X.
   * No, we don't want the EgammaAnalysis code in 90X. I agree with you that this would be useless code repetition. I prefer to have the analysis-level application code just in 80X. If you want, I can port the code recommendation made in this PR to #16968
   * I tested other run1 workflows, and all seem to fail because of the lack of the payloads in the GT... as I mentioned above, I wrote to @mmusich about this issue. When the GTs are update we proceed with the tests here. :)

Thanks.

**Comment by slava77** [comment]
Hi Rafael

On 12/28/16 2:07 PM, Rafael Lopes de Sa wrote:
> Hi @slava77 <https://github.com/slava77>,
> 
> I've updated the codes with your recommendations. Some further comments
> below:
> 
>   * Yes, using the qOverp error returned the same numerical value (to
>     numerical precision). I changed as you recommended.

Nice.

>   * You are correct that the old regression style will /not/ be
>     supported in 90X.
>   * No, we don't want the EgammaAnalysis code in 90X. I agree with you
>     that this would be useless code repetition. I prefer to have the
>     analysis-level application code just in 80X. If you want, I can port
>     the code recommendation made in this PR to #16968
>     <https://github.com/cms-sw/cmssw/pull/16968>

I thought that EGExtraInfoModifierFromDBUser.cc was proposed already for 90X
when I wrote the last message.
Good that it wasn't.

Which part of the recipe did you have in mind to port from (the yet
unmerged) 80X PR?
Is it going to be the code to possibly read (mini)AOD made with the old
regressions in 80X
to modify/update using the new one and use them in 90X?
This doesn't seem like a necessary use case.


>   * I tested other run1 workflows, and all seem to fail because of the
>     lack of the payloads in the GT... as I mentioned above, I wrote to
>     @mmusich <https://github.com/mmusich> about this issue. When the GTs
>     are update we proceed with the tests here. :)

OK

**Comment by rafaellopesdesa** [comment]
Hi Slava,

Answering specifically the new questions:

   * Which part of the recipe did you have in mind to port from (the yet
unmerged) 80X PR?

Not from, but to. I meant recommendations concerning variable names and types, recipe to calculate pModeError, and so on... I am really interested in having the 80X PR merged. Even though it is an analysis-level PR, a lot of people are already applying the energy corrections in their analysis and I think it's much cleaner to have it in the central repository than in my personal area (that it naturally getting more and more outdated with every merge in 80X).

   * Is it going to be the code to possibly read (mini)AOD made with the old
regressions in 80X to modify/update using the new one and use them in 90X?

First, let me second your opinion that I don't think this will be necessary, but...

with this PR, what you said can be done at configuration level in RecoEgamma (I couldn't do this in 80X because the release was already frozen to anything changing RECO). I could provide a configuration suited to re-run (mini)AOD and add the new regression. Even better, I would do that as an option in regressionModifier_cfi.py. 

But, for the meantime, I won't add this option because I honestly think no one will need it at this point.

Thanks.

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 7]
```diff
@@ -253,52 +259,83 @@ void EGExtraInfoModifierFromDB::setEvent(const edm::Event& evt) {
...
-  edm::ESHandle<GBRForest> forestEH;
+  GBRForestD* forest;
+  
+  const std::vector<std::string> ph_condnames_ecalonly_mean  = ph_conf.condnames_ecalonly_mean;
+  const std::vector<std::string> ph_condnames_ecalonly_sigma = ph_conf.condnames_ecalonly_sigma;
 
-  const std::vector<std::string> ph_condnames_mean  = (bunchspacing_ == 25) ? ph_conf.condnames_mean_25ns  : ph_conf.condnames_mean_50ns;
-  const std::vector<std::string> ph_condnames_sigma = (bunchspacing_ == 25) ? ph_conf.condnames_sigma_25ns : ph_conf.condnames_sigma_50ns;
+  if (useLocalFile_)
+    pointerLocalFile_ = TFile::Open(addressLocalFile_.fullPath().c_str(), "READ");
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: file open and close every event is another no-go.
Please move to the constructor or something that's not executed every event.

**Comment by slava77** [comment]
On 12/28/16 3:33 PM, Rafael Lopes de Sa wrote:
> Hi Slava,
> 
> Answering specifically the new questions:
> 
>   * Which part of the recipe did you have in mind to port from (the yet
>     unmerged) 80X PR?
> 
> Not from, but to.
> 

Ah, sorry.


The 80X PR is a sideways "backport" with  different reviewers assigned.
Not particularly close to ideal situation.

It may be more practical that the new
EGExtraInfoModifierFromDB is renamed to EGExtraInfoModifierFromDB2016
(or a similar name).
It will then be kept in RecoEgamma/EgammaTools in both 80X and 90X.
(the old EGExtraInfoModifierFromDB can be removed already in this PR).
The backport to 80X then becomes more straightforward.

**Comment by rafaellopesdesa** [comment]
Hi Slava,

Ok, I updated the code to just open the file once. I also notice that the original code was growing the vector of GBRForestD pointers in every event. Now I just reserve the size I need and update in every event.

Rafael

**Comment by rafaellopesdesa** [comment]
About the suggestion of how to deal with the 80X PR, I think I prefer the file structure as it is right now. I understand that is less ideal for code review, but I think it is easier for us in EGM to find things in the future. As I've explained in the 80X PR, EgammaAnalysis has been traditionally used for user-level applications, and I was hoping to be faithful to this idea.

Let's see how these PR go. Tomorrow I will update the 80X PR with the improvements discussed today in this thread.

**Comment by mmusich** [comment]
@rafaellopesdesa @slava77
about:

>  I tested other run1 workflows, and all seem to fail because of the lack of the payloads in the GT... as I mentioned above, I wrote to @mmusich about this issue. When the GTs are update we proceed with the tests here. :)

the updates are being tested now in #17104. I suggest that commit https://github.com/cms-AlCaDB/cmssw/commit/aee8c69da128c444aca18ce03ee1d8297c94baaf is pulled in together with the rest of the code changes requested during the review to be able to execute tests. 
Cheers,

M.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@rafaellopesdesa 
I was looking at slides in the EGM/ECAL meeting today.
Should I interpret that you will soon update the payloads or/and the variables used in training?
Please clarify.
Thank you.

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 8]
```diff
@@ -253,52 +306,62 @@ void EGExtraInfoModifierFromDB::setEvent(const edm::Event& evt) {
...
 void EGExtraInfoModifierFromDB::setEventContent(const edm::EventSetup& evs) {
 
   iSetup_ = &evs;
 
   edm::ESHandle<GBRForestD> forestDEH;
-  edm::ESHandle<GBRForest> forestEH;
 
-  const std::vector<std::string> ph_condnames_mean  = (bunchspacing_ == 25) ? ph_conf.condnames_mean_25ns  : ph_conf.condnames_mean_50ns;
-  const std::vector<std::string> ph_condnames_sigma = (bunchspacing_ == 25) ? ph_conf.condnames_sigma_25ns : ph_conf.condnames_sigma_50ns;
+  const std::vector<std::string> ph_condnames_ecalonly_mean  = ph_conf.condnames_ecalonly_mean;
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: use ```const std::vector<std::string>& ph_condnames_ecalonly_mean  = ph_conf.condnames_ecalonly_mean;``` to avoid unnecessary copies

Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: same for all other instances in this file

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 9]
```diff
@@ -364,118 +426,128 @@ namespace {
...
+  const bool iseb = ele.isEB();  
+  edm::Handle<EcalRecHitCollection > ecalRecHits;
+  if (iseb) ecalRecHits = ecalRecHitsEB_ ;
+  else      ecalRecHits = ecalRecHitsEE_ ;
+  if ( !ecalRecHits.isValid() ) return;
   
-  std::array<float, 33> eval;  
+  int nSaturatedXtals  = 0;
+  std::vector< std::pair<DetId, float> > hitsAndFractions = theseed->hitsAndFractions();
+  for (auto hitFractionPair : hitsAndFractions) {
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: ```const std::vector< std::pair<DetId, float> >& hitsAndFractions ```
``` for (auto const& hitFractionPair : hitsAndFractions ```

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 10]
```diff
@@ -364,118 +426,128 @@ namespace {
...
-  float maxDR = 0;
   edm::Ptr<reco::CaloCluster> pclus;
   // loop over all clusters that aren't the seed  
-  auto clusend = the_sc->clustersEnd();
-  for( auto clus = the_sc->clustersBegin(); clus != clusend; ++clus ) {
-    pclus = *clus;
-    
-    if(theseed == pclus ) 
+  size_t i_cluster = 0;
+  for( auto clus = the_sc->clustersBegin(); clus != the_sc->clustersEnd(); ++clus ) {
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: why not just ``` for (auto const& pclus: the_sc->clusters() ) ```  ?

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc (hunk 4 above)
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: Please define ``` rawInverse ``` in  a similar way. It's used multiple times here (in [3] and then in [25-27])

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 11]
```diff
@@ -364,118 +426,128 @@ namespace {
...
+    eval[35] = eeseedid.iy();
+    eval[36] = raw_es_energy/raw_energy;
   }
-
+  
   //magic numbers for MINUIT-like transformation of BDT output onto limited range
   //(These should be stored inside the conditions object in the future as well)
-  constexpr double meanlimlow  = 0.2;
-  constexpr double meanlimhigh = 2.0;
+  constexpr double meanlimlow  = -1.0;
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: Looks like the comment about saving these magic numbers in the conditions did not serve the purpose.
Better happen next time.

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 12]
```diff
@@ -484,189 +556,234 @@ void EGExtraInfoModifierFromDB::modifyObject(reco::GsfElectron& ele) const {
...
   if( missing_clusters ) return ; // do not apply corrections in case of missing info (slimmed MiniAOD electrons)
 
+  const bool iseb = pho.isEB();  
+  edm::Handle<EcalRecHitCollection > ecalRecHits;
+  if (iseb) ecalRecHits = ecalRecHitsEB_ ;
+  else      ecalRecHits = ecalRecHitsEE_ ;
+
+  int nSaturatedXtals  = 0;
+  std::vector< std::pair<DetId, float> > hitsAndFractions = theseed->hitsAndFractions();
+  for (auto hitFractionPair : hitsAndFractions) {
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: as above, add ```& ``` to avoid copies

File: RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc [hunk 13]
```diff
@@ -484,189 +556,234 @@ void EGExtraInfoModifierFromDB::modifyObject(reco::GsfElectron& ele) const {
...
+  clusterRawEnergy.resize(std::max(3, numberOfClusters), 0);
+  std::vector<float> clusterDEtaToSeed;
+  clusterDEtaToSeed.resize(std::max(3, numberOfClusters), 0);
+  std::vector<float> clusterDPhiToSeed;
+  clusterDPhiToSeed.resize(std::max(3, numberOfClusters), 0);
+  
+  edm::Ptr<reco::CaloCluster> pclus;
+  // loop over all clusters that aren't the seed  
+  size_t i_cluster = 0;
+  for( auto clus = the_sc->clustersBegin(); clus != the_sc->clustersEnd(); ++clus ) {
```
Comment (slava77) [inline RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc]: as above, range loop can be simpler

**Comment by rafaellopesdesa** [comment]
Hi Slava,

Yes, Fabrice pointed to me an inconsistency between the training code and
application code, so a small update will be introduced in the coming days.
I will also reply to your more recent code review concomitantly.

Thanks again,
-- Rafael.

**Comment by slava77** [comment]
@rafaellopesdesa 
I just wanted to check the progress here.

**Comment by slava77** [comment]
(almost) weekly ping.
@rafaellopesdesa 
Please update on the status of this PR

Thank you.

**Comment by rafaellopesdesa** [comment]
@slava77 Iterating with AlCA to add the payloads and just made an associated PR so that I don't have to re-calculate values when applying the regression.

**Comment by rafaellopesdesa** [comment]
Closing and merging with #17370 with the corrections required in both reviews.
